### DISCUSSION TITLE
Is crypto_in supposed to be the amount before or after the fee is taken?

### DISCUSSION THREAD

**Discussion by orientalperil** [eprbell/rp2#94] [Q&A] [answered]
I am entering my Binance.us transaction data with manual CSVs into DaLI and RP2 but am having some issues understanding how to model the cost basis.

Buy transaction:

```
{
    "symbol": "BTCUSD4",
    "id": "7468480",
    "orderId": "242050613",
    "orderListId": "-1",
    "price": "48233.8200",
    "qty": "0.00414000",
    "quoteQty": "199.6880",
    "commission": "0.00000414",
    "commissionAsset": "BTC",
    "time": "1613441463786",
    "isBuyer": true,
    "isMaker": false,
    "isBestMatch": true
}
```

I tried to enter my buy transaction like this:

```
spot_price: 48233.8200
crypto_in: 0.00413586
crypto_fee: 0.00000414
fiat_in_no_fee: 199.4883
fiat_in_with_fee: 199.6880
```

The `crypto_in` field is described as without fee.  Since I bought 0.00414 with a fee of 10 basis points the amount received should be 0.00414*0.999=0.00413586.  When I use this number as the `crypto_in` field, RP2 creates a fee transaction and deducts the fee from that amount.  This is **wrong** because my balance should be 0.00413586 **not** 0.00413586-0.00000414=0.00413172.

Then I tried to use 0.00414 as `crypto_in` which gives the correct balance, but then I get this warning:

```
WARNING: BTC InTransaction (2021-02-16 02:11:03.786000+00:00, id 3): crypto_in * spot_price != fiat_in_no_fee: 199.688015 != 199.488300
```

On the sell side I'm also having a problem.  It seems RP2 is assuming the fee must be in crypto.

Sell transaction:

```
{
    "symbol": "BTCUSD4",
    "id": "28224293",
    "orderId": "797909901",
    "orderListId": "-1",
    "price": "41484.4400",
    "qty": "0.44500000",
    "quoteQty": "171.5738",
    "commission": "0.1700",
    "commissionAsset": "USD4",
    "time": "1641538837030",
    "isBuyer": false,
    "isMaker": false,
    "isBestMatch": true
}
```

I tried to enter my sell transaction like this:

```
spot_price: 41484.44
crypto_out_no_fee: 0.00413586
crypto_fee: 0
fiat_out_no_fee: 171.4038
fiat_fee: 0.17
```

RP2 is saying that the `crypto_out_no_fee` and `crypto_fee` fields are required.  But Binance.us takes the fee in fiat when selling (`commissionAsset` is `USD4`).  I set `crypto_fee` to zero, but get this warning:

```
WARNING: BTC OutTransaction (2022-01-07 07:00:37.030000+00:00, id 8): crypto_fee * spot_price != fiat_fee: 0.000000 != 0.170000
WARNING: BTC OutTransaction (2022-01-07 07:00:37.030000+00:00, id 8): crypto_out_no_fee * spot_price != fiat_out_no_fee: 171.573836 != 171.403800
```

Questions:

1) Is `crypto_in` supposed to be the amount before or after the fee is taken?
2) For `InTransaction` why does RP2 deduct the fee from a balance that was without fee?
3) How do I model the `OutTransaction` so that the fee comes from fiat?
4) How should I enter these transactions?

**Comment by eprbell** [comment]
> I am entering my Binance.us transaction data with manual CSVs into DaLI and RP2 but am having some issues understanding how to model the cost basis.
> 
> Buy transaction:
> 
> ```
> {
>     "symbol": "BTCUSD4",
>     "id": "7468480",
>     "orderId": "242050613",
>     "orderListId": "-1",
>     "price": "48233.8200",
>     "qty": "0.00414000",
>     "quoteQty": "199.6880",
>     "commission": "0.00000414",
>     "commissionAsset": "BTC",
>     "time": "1613441463786",
>     "isBuyer": true,
>     "isMaker": false,
>     "isBestMatch": true
> }
> ```
> 
> I tried to enter my buy transaction like this:
> 
> ```
> spot_price: 48233.8200
> crypto_in: 0.00413586
> crypto_fee: 0.00000414
> fiat_in_no_fee: 199.4883
> fiat_in_with_fee: 199.6880
> ```
> 
> The `crypto_in` field is described as without fee. Since I bought 0.00414 with a fee of 10 basis points the amount received should be 0.00414*0.999=0.00413586. When I use this number as the `crypto_in` field, RP2 creates a fee transaction and deducts the fee from that amount. This is **wrong** because my balance should be 0.00413586 **not** 0.00413586-0.00000414=0.00413172.
> 
> Then I tried to use 0.00414 as `crypto_in` which gives the correct balance, but then I get this warning:
> 
> ```
> WARNING: BTC InTransaction (2021-02-16 02:11:03.786000+00:00, id 3): crypto_in * spot_price != fiat_in_no_fee: 199.688015 != 199.488300
> ```
> On the sell side I'm also having a problem. It seems RP2 is assuming the fee must be in crypto.
> 
> Sell transaction:
> 
> ```
> {
>     "symbol": "BTCUSD4",
>     "id": "28224293",
>     "orderId": "797909901",
>     "orderListId": "-1",
>     "price": "41484.4400",
>     "qty": "0.44500000",
>     "quoteQty": "171.5738",
>     "commission": "0.1700",
>     "commissionAsset": "USD4",
>     "time": "1641538837030",
>     "isBuyer": false,
>     "isMaker": false,
>     "isBestMatch": true
> }
> ```
> 
> I tried to enter my sell transaction like this:
> 
> ```
> spot_price: 41484.44
> crypto_out_no_fee: 0.00413586
> crypto_fee: 0
> fiat_out_no_fee: 171.4038
> fiat_fee: 0.17
> ```
> 
> RP2 is saying that the `crypto_out_no_fee` and `crypto_fee` fields are required. But Binance.us takes the fee in fiat when selling (`commissionAsset` is `USD4`). I set `crypto_fee` to zero, but get this warning:
> 
> ```
> WARNING: BTC OutTransaction (2022-01-07 07:00:37.030000+00:00, id 8): crypto_fee * spot_price != fiat_fee: 0.000000 != 0.170000
> WARNING: BTC OutTransaction (2022-01-07 07:00:37.030000+00:00, id 8): crypto_out_no_fee * spot_price != fiat_out_no_fee: 171.573836 != 171.403800
> ```
> 
> Questions:
> 
>     1. Is `crypto_in` supposed to be the amount before or after the fee is taken?
>
>     2. For `InTransaction` why does RP2 deduct the fee from a balance that was without fee?
> 
>     3. How do I model the `OutTransaction` so that the fee comes from fiat?
> 
>     4. How should I enter these transactions?

You don't need to define the fiat fields: RP2 will compute them for you internally. They are there only because some exchanges (Coinbase, for example) provides both fiat and crypto values and sometimes the fiat values don't match their expected values. In such case you're supposed to enter both crypto and fiat values (and RP2 will use the passed-in fiat values without computing them), but you'll get a warning similar to the one you pasted. If the exchange doesn't provide fiat values or if it does and they are the correct value, there is no need to enter them.

`crypto_in` is the amount of crypto without the fee (see: https://github.com/eprbell/rp2/blob/main/docs/input_files.md#in-transaction-table-format). As for the output transaction, are you sure the fee is actually fiat? Sometimes exchanges give you the fiat value of the fee, but they actually took the fee out of the crypto stash (check both your crypto and fiat balances on the exchange). All exchanges I have used charge fees in out transactions, but sometimes they show it as fiat. The only time when it would be appropriate to use a fiat out fee is if the exchange took it from your fiat stash.

I cannot guide you exactly on how to enter them because I don't have a binance.us account, but the guiding principles are:
* don't use the fiat fields, unless they are given by the exchange and they don't match their expected values
* know that the out fee is sometimes given in fiat, but it's actually crypto 
* use the balances to understand if you're doing the right thing, both for in and out transactions: the crypto balance generated by RP2 should match your crypto balance on the exchange.

Hope this helps.

**Comment by orientalperil** [comment]
Thanks for the explanation.  It makes a lot more sense about the fiat fields and how fees on output transactions are considered.

Can you clarify more about `crypto_in`.  The way this field works seems contradictory to me.

Working through my example, I bought 0.00414 BTC for $199.6880.  The fee is 10 basis points of the BTC so I enter:

```
spot_price: 48233.82
crypto_in: 0.00413586
crypoto_fee: 0.00000414
fiat_in_no_fee: 199.4883
fiat_in_with_fee: 199.6880
```

I have only this one transaction in my spreadsheet.  RP2 calculates that my balance is 0.00413172, but this is not right.  This is counting the fee twice.  Once when I compute `crypto_in` and then again when it deducts the fee.  If `crypto_in` is the amount of crypto without the fee then why is RP2 deducting the fee from my balance which doesn't have the fee?
![in-out](https://user-images.githubusercontent.com/151680/231343719-06e69b52-1511-4487-ada7-5c1fb984e33f.png)
![tax](https://user-images.githubusercontent.com/151680/231343724-7cdfceda-0033-44b6-b858-cd00d56806e6.png)

**Reply by eprbell** [reply, marked as answer]
I think my initial explanation left out an important detail. For in transactions the fee is usually actually fiat (contrary to out transactions, where it's crypto). In other words when you buy crypto you give the exchange X fiat and the exchange takes a part of X as fee and gives you the equivalent of X - fiat fee, converted to crypto. Whereas for out transactions it's the other way around: typically you sell Y crypto and the exchange takes a part of Y as fee and gives you the equivalent of Y - crypto fee, converted to fiat.

So what you're seeing in RP2's output is correct:
* after the in transaction your balance is 0.00413586 (because the fee was paid in fiat and came out of your fiat stash)
* after the out transaction your balance is 0.00413172 (because the fee was paid in crypto, so 0.00413586 - 0.00414)

Maybe this topic deserves to be covered in the FAQ. There is [already one about fees](https://github.com/eprbell/rp2/blob/main/docs/user_faq.md#how-to-represent-fiat-vs-crypto-transaction-fees), but I think it can be extended.

Hope this helps.

**Comment by eprbell** [comment]
To clarify further, if after your in transaction your crypto balance was 0.00414, then you should enter 0.00414 as `crypto_in`: that doesn't include the fee, which was paid in fiat anyway. Then the rest should fall into place.

**Comment by orientalperil** [comment]
Oh ok I think I get it.  `crypto_in` is 0.00413586 because that's what my balance is.  So for the in transaction I set `crypto_fee` to nothing and `fiat_fee` to the USD equivalent of 0.00000414 at the spot price ($48233.82) which is $0.1996880148.

I got my reports generated with correct cost basis! Thanks! 👍

For an in transaction where `crypto_fee` is defined, RP2 is generating an out transaction for that crypto amount.  Alternatively for an in transaction you can enter `fiat_fee` which will just add the fee to your cost basis.  This seems to be different from out transactions where `crypto_fee` is required and RP2 computes the fiat fee internally (or the user can optionally provide an overriding value).  It makes sense to me that if in transactions can have fees written in either crypto or fiat that out transactions should have a way to do that too.  But maybe it doesn't matter if you can model it with `crypto_fee` as zero and providing `fiat_fee` override.

Also this is interesting because this is backwards from the data that Binance reports in its API.  For in transactions Binance takes the fee in crypto units and for out transactions Binance takes the fee in fiat units.  They have a `commissionAsset` field to indicate this.

It's possible that it could be as you described above where out transactions actually take part of the crypto as a fee and it is just reported in USD.  I can't tell which it is without more evidence.  But the `commissionAsset` field makes me think that they take the commission as they say they do.

**Reply by eprbell** [reply]
You should be able to verify what the exchange actually did by looking at the balances. If they got the fee out of your fiat stash then it's a fiat fee, if they got it out of your crypto stash then it's a crypto fee. What the exchange reports is sometimes misleading: they may report crypto fee in fiat and viceversa, but the balances are a surefire way to find out. Glad you were able to generate the correct report! If you find RP2 and DaLI useful, please consider giving them a star on Github.
